LibreHardwareMonitorLib/PawnIo/PawnIo.cs

```diff
@@ -1,175 +1,122 @@
 
-            return new Version((int)((version >> 16) & 0xFF),
-                               (int)((version >> 8) & 0xFF),
-                               (int)(version & 0xFF),
-                               0);
-        }
-        catch
-        {
-            return new Version();
-        }
-    }
+    private static Lazy<Version> _version = new(() => {
+        using RegistryKey subKey = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\PawnIO");
+        if (subKey?.GetValue("DisplayVersion") is string displayVersion && System.Version.TryParse(displayVersion, out Version version))
+            return version;
 
-    public void Close()
-    {
-        if (_handle != IntPtr.Zero)
-            pawnio_close(_handle);
-    }
+        return null;
+    });
+
+    private readonly SafeFileHandle _handle;
+
+    private PawnIo(SafeFileHandle handle) => _handle = handle;
 
     public static PawnIo LoadModuleFromResource(Assembly assembly, string resourceName)
     {
-        var pawnIO = new PawnIo();
+        SafeFileHandle handle = CreateFile(@"\\.\PawnIO", FileAccess.ReadWrite, FileShare.ReadWrite, IntPtr.Zero, FileMode.Open, FileAttributes.Normal, IntPtr.Zero);
+
+        if (handle.IsInvalid)
+            return new(null);
 
-        using Stream s = assembly.GetManifestResourceStream(resourceName);
+        uint read = 0;
 
-        if (s is UnmanagedMemoryStream ums)
+        byte[] bin = _resourceCache.GetOrAdd((assembly, resourceName), key =>
         {
-            TryLoadLibrary();
-
-            try
-            {
-                pawnio_open(out IntPtr handle);
-                pawnio_load(handle, ums.PositionPointer, (IntPtr)ums.Length);
-                pawnIO._handle = handle;
-            }
-            catch
-            {
-                // PawnIO is not available.
-            }
-        }
+            using var stream = assembly.GetManifestResourceStream(resourceName);
+            using MemoryStream memory = new();
+            stream.CopyTo(memory);
+            return memory.ToArray();
+        });
 
-        return pawnIO;
+        if (DeviceIoControl(handle, ControlCode.LoadBinary, bin, (uint)bin.Length, null, 0u, ref read, IntPtr.Zero))
+            return new(handle);
+
+        return new(null);
     }
 
+    public bool IsLoaded => _handle is
+    {
+        IsInvalid: false,
+        IsClosed: false
+    };
+
+    public void Close() => _handle.Close();
+
     public long[] Execute(string name, long[] input, int outLength)
     {
-        long[] result = new long[outLength];
+        if (_handle is not { IsInvalid: false, IsClosed: false })
+            return [];
+
+        uint read = 0;
 
-        if (_handle == IntPtr.Zero)
-            return result;
+        byte[] output = new byte[outLength * sizeof(long)];
+        byte[] inp = new byte[input.Length * sizeof(long) + 32];
+        Buffer.BlockCopy(Encoding.ASCII.GetBytes(name), 0, inp, 0, Math.Min(31, name.Length));
```

Comment:
Is name always fixed to 32 chars?

Reply:
The first 32 bytes is for the name as null terminated string, so max 31 characters and a null character. The rest is the input array. This is the corresponding [PawnIoLib source](https://github.com/namazso/PawnIOLib/blob/aef64eb8810d93791dd24d8c318db2e3b119b264/PawnIOLib/PawnIOLib.cpp#L90-L93) that I derived this from.